File: C++/copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    unordered_map<Node*, Node*> nodeMap;
    
    Node* copyRandomList(Node* head) {
        Node* ans = copyNodes(head);
        fillRandom(head, ans);
        return ans;
    }
    
    Node* copyNodes(Node* head) {
        if (head) {
            Node* node = new Node(head->val, copyNodes(head->next), nullptr);
            nodeMap[head] = node;
            return node;
        }
        return nullptr;
    }
    
    void fillRandom(Node* src, Node* dst) {
        while (src) {
            if (src->random) {
                dst->random = nodeMap[src->random];
            }
            src = src->next;
            dst = dst->next;
        }
    }
};

File: Kotlin/copy-list-with-random-pointer.kt
/**
 * Example:
 * var ti = Node(5)
 * var v = ti.`val`
 * Definition for a Node.
 * class Node(var `val`: Int) {
 *     var next: Node? = null
 *     var random: Node? = null
 * }
 */

class Solution {
    fun copyRandomList(node: Node?): Node? {
        if (node == null) {
            return null
        }
        val map = mutableMapOf<Int,Node>()
        var curr: Node? = node
        while (curr != null) {
            val newNode = Node(curr.`val`)
            newNode.next = curr.next
            newNode.random = curr.random
            map[curr.hashCode()] = newNode
            curr = curr.next
        }
        var result = map[node.hashCode()]
        curr = result
        while (curr != null) {
            if (curr.next != null)
                curr.next = map[curr.next.hashCode()]
            if (curr.random != null)
                curr.random = map[curr.random.hashCode()]
            curr = curr.next
        }        
        return result
    }
}